nano/node/bootstrap/bootstrap.cpp

```diff
@@ -573,6 +592,159 @@ void nano::bootstrap_attempt::add_bulk_push_target (nano::block_hash const & hea
 	bulk_push_targets.emplace_back (head, end);
 }
 
+void nano::bootstrap_attempt::attempt_restart_check (nano::unique_lock<std::mutex> & lock_a)
+{
+	if (!confirmed_frontiers && requeued_pulls > (!node->network_params.network.is_test_network () ? nano::bootstrap_limits::requeued_pulls_limit : nano::bootstrap_limits::requeued_pulls_limit_test))
+	{
+		confirm_frontiers (lock_a);
+		if (!confirmed_frontiers)
+		{
+			node->stats.inc (nano::stat::type::bootstrap, nano::stat::detail::frontier_confirmation_failed, nano::stat::dir::in);
+			node->bootstrap_initiator.blacklist.add (endpoint_frontier_request);
+			node->logger.always_log (boost::str (boost::format ("Adding peer to blacklist: %1%") % endpoint_frontier_request));
+			for (auto i : clients)
+			{
+				if (auto client = i.lock ())
+				{
+					client->channel->socket->close ();
+				}
+			}
+			idle.clear ();
+			run_start (lock_a);
+		}
+		else
+		{
+			node->stats.inc (nano::stat::type::bootstrap, nano::stat::detail::frontier_confirmation_successful, nano::stat::dir::in);
+		}
+	}
+}
+
+void nano::bootstrap_attempt::confirm_frontiers (nano::unique_lock<std::mutex> & lock_a)
+{
+	// clang-format off
+	condition.wait (lock_a, [& stopped = stopped] { return !stopped; });
+	// clang-format on
+	std::vector<nano::block_hash> frontiers;
+	for (auto i (pulls.begin ()), end (pulls.end ()); i != end && frontiers.size () != nano::bootstrap_limits::bootstrap_max_confirm_frontiers; ++i)
+	{
+		if (!i->head.is_zero ())
+		{
+			frontiers.push_back (i->head);
+		}
+	}
+	for (auto i (recent_pulls_head.begin ()), end (recent_pulls_head.end ()); i != end && frontiers.size () != nano::bootstrap_limits::bootstrap_max_confirm_frontiers; ++i)
+	{
+		if (!(*i).is_zero ())
+		{
+			frontiers.push_back (*i);
+		}
+	}
+	lock_a.unlock ();
+	auto frontiers_count (frontiers.size ());
+	const size_t reps_limit = 20;
+	auto representatives (node->rep_crawler.representatives ());
+	auto reps_weight (node->rep_crawler.total_weight ());
+	auto representatives_copy (representatives);
+	nano::uint128_t total_weight (0);
+	// Select random peers from bottom 50% of principal representatives
+	if (!representatives.empty ())
+	{
+		if (representatives.size () > 1)
+		{
+			std::reverse (representatives.begin (), representatives.end ());
+			representatives.resize (representatives.size () / 2, nano::representative{ 0, 0, nullptr });
```

Comment:
I think it may be simpler to add a default `= default` constructor in `class representative` instead of adding the second argument to all the `resize` calls. The container is only getting smaller in these cases so shouldn't really need them, but I guess the compiler complains. It would reduce some noise, however only it if you think having a default constructor on `representative` is correct.

I did wonder if there could be some optimization on the `reverse` as you only really need to reverse the top half of the representatives as the rest is discarded, but I don't think it's worth the hassle as the container will be relatively small.

Reply:
Updated